### ISSUE TITLE
Get resourcesServerFilter from ResourceServerSecurityConfigurer

### ISSUE THREAD

**Issue by ggeorgijevic** [spring-projects/spring-security-oauth#1189]
I'm in the process of transfering xml to Java configuration for oauth2 resource server.  By extending ResourceServerConfigurerAdapter and using @EnableResourceServer, a resourcesServerFilter (OAuth2AuthenticationProcessingFilter) is automatically added to Spring Security filter chain (at the position before AbstractPreAuthenticatedProcessingFilter). 

I want to add that filter into my filter chain, and then place that filter chain to Spring Security filter chain before AbstractPreAuthenticatedProcessingFilter.

I don't see the way to get the instance of OAuth2AuthenticationProcessingFilter defined in ResourceServerSecurityConfigurer. I tried to create an instance of OAuth2AuthenticationProcessingFilter and it worked well, but that way there are two instances of OAuth2AuthenticationProcessingFilter (the one that I created and the one that's automatically created) and the request goes through the same doFilter method twice. I also tried to remove @EnableResourceServer annotation (to skip automatic creation of the OAuth2AuthenticationProcessingFilter), but then I lost lots of other defaults. 

Is it possible to fetch this resourcesServerFilter in Java config?

In xml config, there is something like  <oauth:resource-server id="resourceServerFilter" ... />  and I could fetch resource server filter by its id and place it in my filter chain without problem.

**Comment by dsyer** [comment]
The `OAuth2AuthenticationProcessingFilter` is not a bean in the `@EnableResourceServer` config model. You could create one yourself, as you suggest. I don't see any other way to do it, but I don't really understand why you would need to either.

**Comment by ggeorgijevic** [comment]
I want to place one custom filter before OAuth2AuthenticationProcessingFilter and one custom filter after it. In xml, I had the following filter chain at the position before PRE_AUTH_FILTER

<beans:bean id="oauthResourceServerFilterChain"
		class="org.springframework.security.web.FilterChainProxy">
		<security:filter-chain-map path-type="ant">
			<security:filter-chain pattern="/**"
				filters="
					filterBeforeResourceServerFilter,
					resourceServerFilter,
					filterAfterResourceServerFilter
					" />
		</security:filter-chain-map>
	</beans:bean>

The one before is responsible for setting tenant context (since I have multi tenancy environment) and the  one after is responsible for some additional checking dependent on client context. So it is important to have that exact order of filters.

Can control that order with the auto instantiated OAuth2AuthenticationProcessingFilter?

**Comment by dsyer** [comment]
> Can I control that order with the auto instantiated OAuth2AuthenticationProcessingFilter?

You can access the `HttpSecurity` configurer (which is the equivalent of `<http.../>` in XML) via a `ResourceServerConfigurer`. That can be used to add custom filters, so I guess that's what you need to do. It doesn't have an API for accessing the existing filters, but you can probably assume that the `OAuth2AuthenticationProcessingFilter` is there already.

**Comment by ggeorgijevic** [comment]
Yes, I already do that and add custom filters in configure(HttpSecurity http) method.. But ResourceServerSecurityConfigurer which instantiates OAuth2AuthenticationProcessingFilter is called after my extension of ResourceServerConfigurer. So I get the exception that OAuth2AuthenticationProcessingFilter is not present in the security filter chain so the custom filters cannot be added after/before it..
